pilot/pkg/proxy/agent.go

```diff
@@ -278,10 +301,76 @@ func (a *agent) reconcile() {
 }
 
 // waitForExit runs the start-up command as a go routine and waits for it to finish
-func (a *agent) waitForExit(config interface{}, epoch int, abortCh <-chan error) {
+func (a *agent) waitForExit(config interface{}, epoch int, abort <-chan error) {
 	log.Infof("Epoch %d starting", epoch)
-	err := a.proxy.Run(config, epoch, abortCh)
-	a.statusCh <- exitStatus{epoch: epoch, err: err}
+	cmd, err := a.proxy.Run(config, epoch, abort)
```

Comment:
@mandarjog this is the bulk of the change - code that used to be inside the proxy.Run impl has been moved up to the agent so the agent can interrogate the exit code, signo, etc.   proxy.Run impl is now mostly responsible for setting up the execution environment.